Zend/zend_closures.c

```diff
@@ -381,7 +381,39 @@ static ZEND_COLD zend_function *zend_closure_get_constructor(zend_object *object
 static int zend_closure_compare(zval *o1, zval *o2) /* {{{ */
 {
 	ZEND_COMPARE_OBJECTS_FALLBACK(o1, o2);
-	return Z_OBJ_P(o1) != Z_OBJ_P(o2);
+
+	zend_closure *lhs = (zend_closure*) Z_OBJ_P(o1);
+	zend_closure *rhs = (zend_closure*) Z_OBJ_P(o2);
+
+	if (!((lhs->func.common.fn_flags & ZEND_ACC_FAKE_CLOSURE) && (rhs->func.common.fn_flags & ZEND_ACC_FAKE_CLOSURE))) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (Z_TYPE(lhs->this_ptr) != Z_TYPE(rhs->this_ptr)) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (Z_TYPE(lhs->this_ptr) == IS_OBJECT && Z_OBJ(lhs->this_ptr) != Z_OBJ(rhs->this_ptr)) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (lhs->called_scope != rhs->called_scope) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (lhs->func.type != rhs->func.type) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (lhs->func.common.scope != rhs->func.common.scope) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (!zend_string_equals(lhs->func.common.function_name, rhs->func.common.function_name)) {
```

Comment:
Should this account for static variable `use` in function/fn? For example, if there's at least one variable being used, then always return false.

Even if it shouldn't, there should be a test of the expected behavior

```php
php > $add = static fn (int $a) => static fn(int $b) => $a + $b;
php > $add1 = $add(1);
php > $add2 = $add(2);
php > var_dump($add1(10));
int(11)

// Should $add1 == $add2?
```

It looks like `use` uses ZEND_BIND_LEXICAL and zend_closure_bind_var_ex, which add pointers to the HashTable static_variables?

i.e. check if the hash has at least one element ?

```
void zend_closure_bind_var_ex(zval *closure_zv, uint32_t offset, zval *val) /* {{{ */
{
	zend_closure *closure = (zend_closure *) Z_OBJ_P(closure_zv);
	HashTable *static_variables = ZEND_MAP_PTR_GET(closure->func.op_array.static_variables_ptr);
	zval *var = (zval*)((char*)static_variables->arData + offset);
	zval_ptr_dtor(var);
	ZVAL_COPY_VALUE(var, val);
}
```

Reply:
This change is for fake closures.

Reply:
Never mind, looks good to me.

I was mistaken, the implementation of `==` that calls zend_class_entry->compare always checks if objects are identical first, so my suggestion was redundant.

For something that's already a closure, first-class callables and Closure::fromCallable return the unmodified closure argument, so the behavior in this PR should be fine.

```c
ZEND_API int ZEND_FASTCALL zend_compare_objects(zval *o1, zval *o2) /* {{{ */
{
	if (Z_OBJ_P(o1) == Z_OBJ_P(o2)) {
		return 0;
	}

	if (Z_OBJ_HT_P(o1)->compare == NULL) {
		return 1;
	} else {
		return Z_OBJ_HT_P(o1)->compare(o1, o2);
	}
}
/* }}} */
```